[WIP] Add support for CoffeMaker to WeMo Binding

See https://community.openhab.org/t/wemo-coffe-maker/12667/8 and
https://github.com/openhab/openhab1-addons/issues/4275
Testing JAR can be found [here](https://drive.google.com/open?id=0BxdTuQI5IDvwTFppUUUzY0E0REk)

Here's what I now see in the file:

`2017-05-01 11:45:38.625 [ERROR] [home.core.thing.binding.ThingFactory] - Thing factory (class org.eclipse.smarthome.binding.wemo.internal.WemoHandlerFactory) returned null on create thing when it reports to support the thing type (wemo:coffee).
2017-05-01 11:45:38.625 [WARN ] [g.discovery.internal.PersistentInbox] - Cannot create thing. No binding found that supports creating a thing of type wemo:coffee.`

--- a/extensions/binding/org.eclipse.smarthome.binding.wemo/ESH-INF/thing/thing-types.xml
+++ b/extensions/binding/org.eclipse.smarthome.binding.wemo/ESH-INF/thing/thing-types.xml
@@ -142,9 +142,7 @@
             <channel id="waterLevelReached" typeId="waterLevelReached"/>
             <channel id="cleanAdvise" typeId="cleanAdvise"/>
             <channel id="filterAdvise" typeId="filterAdvise"/>
-            <channel id="brewing" typeId="brewing"/>
             <channel id="brewed" typeId="brewed"/>
-            <channel id="cleaning" typeId="cleaning"/>
             <channel id="lastCleaned" typeId="lastCleaned"/>
         </channels>
 
@@ -286,6 +284,19 @@
         <item-type>String</item-type>
         <label>Mode</label>
         <description>Shows the operation mode of a WeMo Coffee Maker</description>
+        <state readOnly="true">
+           <options>
+              <option value="Refill">Not Ready: Refill Water and Replace Carafe</option>
+              <option value="PlaceCarafe">Not Ready: Replace Carafe</option>
+              <option value="RefillWater">Not Ready: Refill Water</option>
+              <option value="Ready">Ready</option>
+              <option value="Brewing">Brewing</option>
+              <option value="Brewed">Brewing Finished</option>
+              <option value="CleaningBrewing">Cleaning Coffee Maker</option>
+              <option value="CleaningSoaking">Cleaning Filter</option>
+              <option value="BrewFailCarafeRemoved">Brewing Failed: Carafe Removed</option>
+          </options>
+        </state>
     </channel-type>
 
      <channel-type id="modeTime">
@@ -321,27 +332,14 @@
         <state readOnly="true"></state>
     </channel-type>
 
-     <channel-type id="brewing" advanced="true">
-        <item-type>DateTime</item-type>
-        <label>Brewing</label>
-        <description>This channel is not documented right now. If You find a valid description, please post to community</description>
-    </channel-type>
-
      <channel-type id="brewed" advanced="true">
         <item-type>DateTime</item-type>
         <label>Brewed</label>
         <description>Date/time the coffee maker last completed brewing coffee</description>
     </channel-type>
 
-     <channel-type id="cleaning" advanced="true">
-        <item-type>Number</item-type>
-        <label>Cleaning</label>
-        <description>This channel is not documented right now. If You find a valid description, please post to community</description>
-        <state readOnly="true"></state>
-    </channel-type>
-
      <channel-type id="lastCleaned" advanced="true">
-        <item-type>Number</item-type>
+        <item-type>DateTime</item-type>
         <label>LastCleaned</label>
         <description>Date/time the coffee maker last completed cleaning</description>
     </channel-type>

--- a/extensions/binding/org.eclipse.smarthome.binding.wemo/ESH-INF/thing/thing-types.xml
+++ b/extensions/binding/org.eclipse.smarthome.binding.wemo/ESH-INF/thing/thing-types.xml
@@ -130,6 +130,39 @@
          </config-description>
     </thing-type>
 
+     <thing-type id="CoffeeMaker">
+        <label>Mr. Coffee WeMo Enabled Coffeemaker</label>
+        <description>This is a WeMo enabled coffee maker</description>
+
+        <channels>
+            <channel id="state" typeId="state" />
+            <channel id="coffeeMode" typeId="coffeeMode"/>
+            <channel id="modeTime" typeId="modeTime"/>
+            <channel id="timeRemaining" typeId="timeRemaining"/>
+            <channel id="waterLevelReached" typeId="waterLevelReached"/>
+            <channel id="cleanAdvise" typeId="cleanAdvise"/>
+            <channel id="filterAdvise" typeId="filterAdvise"/>
+            <channel id="brewing" typeId="brewing"/>
+            <channel id="brewed" typeId="brewed"/>
+            <channel id="cleaning" typeId="cleaning"/>
+            <channel id="lastCleaned" typeId="lastCleaned"/>
+        </channels>
+
+        <config-description>
+            <parameter name="udn" type="text">
+                <label>Unique Device Name</label>
+                <description>The UDN identifies the WeMo Device</description>
+                <required>true</required>
+            </parameter>
+            <parameter name="pollingInterval" type="integer" required="false" min="15" max="180">
+                <label>Polling Interval</label>
+                <description>Interval polling the WeMo Coffee Maker.
+                </description>
+                <default>60</default>
+            </parameter>
+         </config-description>
+    </thing-type>
+
      <channel-type id="state">
         <item-type>Switch</item-type>
         <label>Switch</label>
@@ -247,6 +280,70 @@
         <label>Sensor</label>
         <description>Shows the state of the integrated sensor</description>
         <category>Switch</category>
+    </channel-type>
+
+     <channel-type id="coffeeMode">
+        <item-type>String</item-type>
+        <label>Mode</label>
+        <description>Shows the operation mode of a WeMo Coffee Maker</description>
+    </channel-type>
+
+     <channel-type id="modeTime">
+        <item-type>Number</item-type>
+        <label>ModeTime</label>
+        <description>Shows the current amount of time, in minutes, that the Coffee Maker has been in the current mode</description>
+    </channel-type>
+
+     <channel-type id="timeRemaining">
+        <item-type>Number</item-type>
+        <label>TimeRemaining</label>
+        <description>Shows the remaining brewing time of a WeMo Coffee Maker</description>
+    </channel-type>
+
+     <channel-type id="waterLevelReached" advanced="true">
+        <item-type>Switch</item-type>
+        <label>WaterLevelReached</label>
+        <description>Indicates if the WeMo Coffee Maker needs to be refilled</description>
+        <state readOnly="true"></state>
+    </channel-type>
+
+     <channel-type id="cleanAdvise" advanced="true">
+        <item-type>Switch</item-type>
+        <label>CleanAdvise</label>
+        <description>Indicates if a WeMo Coffee Maker needs to be cleaned</description>
+        <state readOnly="true"></state>
+    </channel-type>
+
+     <channel-type id="filterAdvise" advanced="true">
+        <item-type>Switch</item-type>
+        <label>FilterAdvise</label>
+        <description>Indicates if a WeMo Coffee Maker needs to have the filter changed</description>
+        <state readOnly="true"></state>
+    </channel-type>
+
+     <channel-type id="brewing" advanced="true">
+        <item-type>DateTime</item-type>
+        <label>Brewing</label>
+        <description>tba</description>
+    </channel-type>
+
+     <channel-type id="brewed" advanced="true">
+        <item-type>DateTime</item-type>
+        <label>Brewed</label>
+        <description>Date/time the coffee maker last completed brewing coffee</description>
+    </channel-type>
+
+     <channel-type id="cleaning" advanced="true">
+        <item-type>Number</item-type>
+        <label>Cleaning</label>
+        <description>tba</description>
+        <state readOnly="true"></state>
+    </channel-type>
+
+     <channel-type id="lastCleaned" advanced="true">
+        <item-type>Number</item-type>
+        <label>LastCleaned</label>
+        <description>tba</description>
     </channel-type>
  
 </thing:thing-descriptions>

--- a/extensions/binding/org.eclipse.smarthome.binding.wemo/ESH-INF/thing/thing-types.xml
+++ b/extensions/binding/org.eclipse.smarthome.binding.wemo/ESH-INF/thing/thing-types.xml
@@ -324,7 +324,7 @@
      <channel-type id="brewing" advanced="true">
         <item-type>DateTime</item-type>
         <label>Brewing</label>
-        <description>tba</description>
+        <description>This channel is not documented right now. If You find a valid description, please post to community</description>
     </channel-type>
 
      <channel-type id="brewed" advanced="true">
@@ -336,14 +336,14 @@
      <channel-type id="cleaning" advanced="true">
         <item-type>Number</item-type>
         <label>Cleaning</label>
-        <description>tba</description>
+        <description>This channel is not documented right now. If You find a valid description, please post to community</description>
         <state readOnly="true"></state>
     </channel-type>
 
      <channel-type id="lastCleaned" advanced="true">
         <item-type>Number</item-type>
         <label>LastCleaned</label>
-        <description>tba</description>
+        <description>Date/time the coffee maker last completed cleaning</description>
     </channel-type>
  
 </thing:thing-descriptions>

--- a/extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/WemoBindingConstants.java
+++ b/extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/WemoBindingConstants.java
@@ -57,9 +57,7 @@ public class WemoBindingConstants {
     public final static String CHANNEL_WATERLEVELREACHED = "waterLevelReached";
     public final static String CHANNEL_CLEANADVISE = "cleanAdvise";
     public final static String CHANNEL_FILTERADVISE = "filterAdvise";
-    public final static String CHANNEL_BREWING = "brewing";
     public final static String CHANNEL_BREWED = "brewed";
-    public final static String CHANNEL_CLEANING = "cleaning";
     public final static String CHANNEL_LASTCLEANED = "lastCleaned";
 
     // List of thing configuration properties

--- a/extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/WemoBindingConstants.java
+++ b/extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/WemoBindingConstants.java
@@ -31,6 +31,7 @@ public class WemoBindingConstants {
     public final static ThingTypeUID THING_TYPE_BRIDGE = new ThingTypeUID(BINDING_ID, "bridge");
     public final static ThingTypeUID THING_TYPE_MZ100 = new ThingTypeUID(BINDING_ID, "MZ100");
     public final static ThingTypeUID THING_TYPE_MAKER = new ThingTypeUID(BINDING_ID, "Maker");
+    public final static ThingTypeUID THING_TYPE_COFFEE = new ThingTypeUID(BINDING_ID, "CoffeeMaker");
 
     // List of all Channel ids
     public final static String CHANNEL_STATE = "state";
@@ -50,9 +51,21 @@ public class WemoBindingConstants {
     public final static String CHANNEL_RELAY = "relay";
     public final static String CHANNEL_SENSOR = "sensor";
 
+    public final static String CHANNEL_COFFEEMODE = "coffeeMode";
+    public final static String CHANNEL_MODETIME = "modeTime";
+    public final static String CHANNEL_TIMEREMAINING = "timeRemaining";
+    public final static String CHANNEL_WATERLEVELREACHED = "waterLevelReached";
+    public final static String CHANNEL_CLEANADVISE = "cleanAdvise";
+    public final static String CHANNEL_FILTERADVISE = "filterAdvise";
+    public final static String CHANNEL_BREWING = "brewing";
+    public final static String CHANNEL_BREWED = "brewed";
+    public final static String CHANNEL_CLEANING = "cleaning";
+    public final static String CHANNEL_LASTCLEANED = "lastCleaned";
+
     // List of thing configuration properties
     public static final String UDN = "udn";
     public static final String DEVICE_ID = "deviceID";
+    public static final String POLLINGINTERVALL = "pollingInterval";
 
     public final static Set<ThingTypeUID> SUPPORTED_BRIDGE_THING_TYPES = ImmutableSet.of(THING_TYPE_BRIDGE);
 
@@ -62,6 +75,7 @@ public class WemoBindingConstants {
             THING_TYPE_INSIGHT, THING_TYPE_LIGHTSWITCH, THING_TYPE_MOTION);
 
     public final static Set<ThingTypeUID> SUPPORTED_THING_TYPES = ImmutableSet.of(THING_TYPE_SOCKET, THING_TYPE_INSIGHT,
-            THING_TYPE_LIGHTSWITCH, THING_TYPE_MOTION, THING_TYPE_BRIDGE, THING_TYPE_MZ100, THING_TYPE_MAKER);
+            THING_TYPE_LIGHTSWITCH, THING_TYPE_MOTION, THING_TYPE_BRIDGE, THING_TYPE_MZ100, THING_TYPE_MAKER,
+            THING_TYPE_COFFEE);
 
 }

--- a/extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/discovery/WemoDiscoveryParticipant.java
+++ b/extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/discovery/WemoDiscoveryParticipant.java
@@ -109,6 +109,11 @@ public class WemoDiscoveryParticipant implements UpnpDiscoveryParticipant {
                                     device.getIdentity().getUdn().getIdentifierString());
                             return new ThingUID(THING_TYPE_MAKER, device.getIdentity().getUdn().getIdentifierString());
                         }
+                        if (device.getDetails().getModelDetails().getModelName().toLowerCase().startsWith("coffee")) {
+                            logger.debug("Discovered a WeMo Coffe Maker thing with UDN '{}'",
+                                    device.getIdentity().getUdn().getIdentifierString());
+                            return new ThingUID(THING_TYPE_COFFEE, device.getIdentity().getUdn().getIdentifierString());
+                        }
                     }
                 }
             }

--- a/extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoCoffeeHandler.java
+++ b/extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoCoffeeHandler.java
@@ -1,0 +1,522 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.binding.wemo.handler;
+
+import static org.eclipse.smarthome.binding.wemo.WemoBindingConstants.*;
+
+import java.io.StringReader;
+import java.math.BigDecimal;
+import java.net.URL;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.GregorianCalendar;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+
+import org.apache.commons.lang.StringEscapeUtils;
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.smarthome.binding.wemo.internal.http.WemoHttpCall;
+import org.eclipse.smarthome.config.core.Configuration;
+import org.eclipse.smarthome.config.discovery.DiscoveryListener;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.library.types.DateTimeType;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.io.transport.upnp.UpnpIOParticipant;
+import org.eclipse.smarthome.io.transport.upnp.UpnpIOService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.w3c.dom.CharacterData;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.InputSource;
+
+import com.google.common.collect.Sets;
+
+/**
+ * The {@link WemoCoffeeHandler} is responsible for handling commands, which are
+ * sent to one of the channels and to update their states.
+ *
+ * @author Hans-Jörg Merk - Initial contribution
+ */
+
+public class WemoCoffeeHandler extends BaseThingHandler implements UpnpIOParticipant, DiscoveryListener {
+
+    private final Logger logger = LoggerFactory.getLogger(WemoCoffeeHandler.class);
+
+    public final static Set<ThingTypeUID> SUPPORTED_THING_TYPES = Sets.newHashSet(THING_TYPE_COFFEE);
+
+    private Map<String, Boolean> subscriptionState = new HashMap<String, Boolean>();
+
+    private Map<String, String> stateMap = Collections.synchronizedMap(new HashMap<String, String>());
+
+    protected final static int SUBSCRIPTION_DURATION = 600;
+
+    private UpnpIOService service;
+
+    /**
+     * The default refresh interval in Seconds.
+     */
+    private int REFRESH_INTERVAL = 60;
+
+    private ScheduledFuture<?> refreshJob;
+
+    private Runnable refreshRunnable = new Runnable() {
+
+        @Override
+        public void run() {
+            try {
+                if (!isUpnpDeviceRegistered()) {
+                    logger.debug("WeMo UPnP device {} not yet registered", getUDN());
+                }
+
+                updateWemoState();
+                onSubscription();
+
+            } catch (Exception e) {
+                logger.debug("Exception during poll : {}", e);
+            }
+        }
+    };
+
+    public WemoCoffeeHandler(Thing thing, UpnpIOService upnpIOService) {
+
+        super(thing);
+
+        logger.debug("Creating a WemoCoffeeHandler V0.3 for thing '{}'", getThing().getUID());
+
+        if (upnpIOService != null) {
+            this.service = upnpIOService;
+        } else {
+            logger.debug("upnpIOService not set.");
+        }
+
+    }
+
+    @Override
+    public void initialize() {
+
+        Configuration configuration = getConfig();
+
+        if (configuration.get("udn") != null) {
+            logger.debug("Initializing WemoCoffeeHandler for UDN '{}'", configuration.get("udn"));
+            onSubscription();
+            onUpdate();
+            super.initialize();
+        } else {
+            logger.debug("Cannot initalize WemoCoffeeHandler. UDN not set.");
+        }
+
+    }
+
+    @Override
+    public void thingDiscovered(DiscoveryService source, DiscoveryResult result) {
+        if (result.getThingUID().equals(this.getThing().getUID())) {
+            if (getThing().getConfiguration().get(UDN).equals(result.getProperties().get(UDN))) {
+                logger.trace("Discovered UDN '{}' for thing '{}'", result.getProperties().get(UDN),
+                        getThing().getUID());
+                updateStatus(ThingStatus.ONLINE);
+                onSubscription();
+                onUpdate();
+            }
+        }
+    }
+
+    @Override
+    public void thingRemoved(DiscoveryService source, ThingUID thingUID) {
+        if (thingUID.equals(this.getThing().getUID())) {
+            logger.trace("Setting status for thing '{}' to OFFLINE", getThing().getUID());
+            updateStatus(ThingStatus.OFFLINE);
+        }
+    }
+
+    @Override
+    public void dispose() {
+        logger.debug("WeMoCoffeeHandler disposed.");
+
+        removeSubscription();
+
+        if (refreshJob != null && !refreshJob.isCancelled()) {
+            refreshJob.cancel(true);
+            refreshJob = null;
+        }
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.trace("Command '{}' received for channel '{}'", command, channelUID);
+
+        if (command instanceof RefreshType) {
+            try {
+                updateWemoState();
+            } catch (Exception e) {
+                logger.debug("Exception during poll : {}", e);
+            }
+        } else if (channelUID.getId().equals(CHANNEL_STATE)) {
+
+            if (command instanceof OnOffType) {
+
+                if (command.equals(OnOffType.ON)) {
+
+                    try {
+
+                        String soapHeader = "\"urn:Belkin:service:deviceevent:1#SetAttributes\"";
+
+                        String content = "<?xml version=\"1.0\"?>"
+                                + "<s:Envelope xmlns:s=\"http://schemas.xmlsoap.org/soap/envelope/\" s:encodingStyle=\"http://schemas.xmlsoap.org/soap/encoding/\">"
+                                + "<s:Body>" + "<u:SetAttributes xmlns:u=\"urn:Belkin:service:deviceevent:1\">"
+                                + "<attributeList>&lt;attribute&gt;&lt;name&gt;Brewed&lt;/name&gt;&lt;value&gt;NULL&lt;/value&gt;&lt;/attribute&gt;"
+                                + "&lt;attribute&gt;&lt;name&gt;LastCleaned&lt;/name&gt;&lt;value&gt;NULL&lt;/value&gt;&lt;/attribute&gt;&lt;attribute&gt;"
+                                + "&lt;name&gt;ModeTime&lt;/name&gt;&lt;value&gt;NULL&lt;/value&gt;&lt;/attribute&gt;&lt;attribute&gt;&lt;name&gt;Brewing&lt;/name&gt;"
+                                + "&lt;value&gt;NULL&lt;/value&gt;&lt;/attribute&gt;&lt;attribute&gt;&lt;name&gt;TimeRemaining&lt;/name&gt;&lt;value&gt;NULL&lt;/value&gt;"
+                                + "&lt;/attribute&gt;&lt;attribute&gt;&lt;name&gt;WaterLevelReached&lt;/name&gt;&lt;value&gt;NULL&lt;/value&gt;&lt;/attribute&gt;&lt;"
+                                + "attribute&gt;&lt;name&gt;Mode&lt;/name&gt;&lt;value&gt;4&lt;/value&gt;&lt;/attribute&gt;&lt;attribute&gt;&lt;name&gt;CleanAdvise&lt;/name&gt;"
+                                + "&lt;value&gt;NULL&lt;/value&gt;&lt;/attribute&gt;&lt;attribute&gt;&lt;name&gt;FilterAdvise&lt;/name&gt;&lt;value&gt;NULL&lt;/value&gt;&lt;/attribute&gt;"
+                                + "&lt;attribute&gt;&lt;name&gt;Cleaning&lt;/name&gt;&lt;value&gt;NULL&lt;/value&gt;&lt;/attribute&gt;</attributeList>"
+                                + "</u:SetAttributes>" + "</s:Body>" + "</s:Envelope>";
+
+                        String wemoURL = getWemoURL("deviceevent");
+
+                        if (wemoURL != null) {
+                            String wemoCallResponse = WemoHttpCall.executeCall(wemoURL, soapHeader, content);
+                            if (wemoCallResponse != null) {
+                                updateState(CHANNEL_STATE, OnOffType.ON);
+                                State newMode = new StringType("Brewing");
+                                updateState(CHANNEL_COFFEEMODE, newMode);
+                            }
+                        }
+                    } catch (Exception e) {
+                        logger.error("Failed to send command '{}' for device '{}': {}", command, getThing().getUID(),
+                                e.getMessage());
+                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+                    }
+
+                } else if (command.equals(OnOffType.OFF)) {
+                    // do nothing, as WeMo Coffee Maker cannot be switched off remotely
+                }
+                updateStatus(ThingStatus.ONLINE);
+            }
+        }
+    }
+
+    @Override
+    public void onServiceSubscribed(String service, boolean succeeded) {
+        logger.debug("WeMo {}: Subscription to service {} {}", getUDN(), service, succeeded ? "succeeded" : "failed");
+        subscriptionState.put(service, succeeded);
+    }
+
+    @Override
+    public void onValueReceived(String variable, String value, String service) {
+        // We can subscribe to GENA events, but there is no usefull response right now.
+    }
+
+    private synchronized void onSubscription() {
+        if (service.isRegistered(this)) {
+            logger.debug("Checking WeMo GENA subscription for '{}'", this);
+
+            String subscription = "deviceevent1";
+            if ((subscriptionState.get(subscription) == null) || !subscriptionState.get(subscription).booleanValue()) {
+                logger.debug("Setting up GENA subscription {}: Subscribing to service {}...", getUDN(), subscription);
+                service.addSubscription(this, subscription, SUBSCRIPTION_DURATION);
+                subscriptionState.put(subscription, true);
+            }
+
+        } else {
+            logger.debug("Setting up WeMo GENA subscription for '{}' FAILED - service.isRegistered(this) is FALSE",
+                    this);
+        }
+    }
+
+    private synchronized void removeSubscription() {
+        logger.debug("Removing WeMo GENA subscription for '{}'", this);
+
+        if (service.isRegistered(this)) {
+
+            String subscription = "deviceevent1";
+            if ((subscriptionState.get(subscription) != null) && subscriptionState.get(subscription).booleanValue()) {
+                logger.debug("WeMo {}: Unsubscribing from service {}...", getUDN(), subscription);
+                service.removeSubscription(this, subscription);
+            }
+
+            subscriptionState = new HashMap<String, Boolean>();
+            service.unregisterParticipant(this);
+        }
+    }
+
+    private synchronized void onUpdate() {
+        if (refreshJob == null || refreshJob.isCancelled()) {
+            Configuration config = getThing().getConfiguration();
+            int refreshInterval = REFRESH_INTERVAL;
+            Object refreshConfig = config.get("pollingInterval");
+            if (refreshConfig != null) {
+                refreshInterval = ((BigDecimal) refreshConfig).intValue();
+                logger.debug("Setting WemoCoffeeHandler refreshInterval to '{}' seconds", refreshInterval);
+            }
+            refreshJob = scheduler.scheduleAtFixedRate(refreshRunnable, 0, refreshInterval, TimeUnit.SECONDS);
+        }
+    }
+
+    private boolean isUpnpDeviceRegistered() {
+        return service.isRegistered(this);
+    }
+
+    @Override
+    public String getUDN() {
+        return (String) this.getThing().getConfiguration().get(UDN);
+    }
+
+    /**
+     * The {@link updateWemoState} polls the actual state of a WeMo CoffeeMaker.
+     */
+    protected void updateWemoState() {
+
+        String action = "GetAttributes";
+        String actionService = "deviceevent";
+
+        String soapHeader = "\"urn:Belkin:service:" + actionService + ":1#" + action + "\"";
+        String content = "<?xml version=\"1.0\"?>"
+                + "<s:Envelope xmlns:s=\"http://schemas.xmlsoap.org/soap/envelope/\" s:encodingStyle=\"http://schemas.xmlsoap.org/soap/encoding/\">"
+                + "<s:Body>" + "<u:" + action + " xmlns:u=\"urn:Belkin:service:" + actionService + ":1\">" + "</u:"
+                + action + ">" + "</s:Body>" + "</s:Envelope>";
+
+        try {
+            String wemoURL = getWemoURL(actionService);
+            if (wemoURL != null) {
+                String wemoCallResponse = WemoHttpCall.executeCall(wemoURL, soapHeader, content);
+                if (wemoCallResponse != null) {
+                    try {
+                        String stringParser = StringUtils.substringBetween(wemoCallResponse, "<attributeList>",
+                                "</attributeList>");
+
+                        // Due to Belkins bad response formatting, we need to run this twice.
+                        stringParser = StringEscapeUtils.unescapeXml(stringParser);
+                        stringParser = StringEscapeUtils.unescapeXml(stringParser);
+
+                        logger.trace("CoffeeMaker response '{}' for device '{}' received", stringParser,
+                                getThing().getUID());
+
+                        stringParser = "<data>" + stringParser + "</data>";
+
+                        DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+                        DocumentBuilder db = dbf.newDocumentBuilder();
+                        InputSource is = new InputSource();
+                        is.setCharacterStream(new StringReader(stringParser));
+
+                        Document doc = db.parse(is);
+                        NodeList nodes = doc.getElementsByTagName("attribute");
+
+                        // iterate the attributes
+                        for (int i = 0; i < nodes.getLength(); i++) {
+                            Element element = (Element) nodes.item(i);
+
+                            NodeList deviceIndex = element.getElementsByTagName("name");
+                            Element line = (Element) deviceIndex.item(0);
+                            String attributeName = getCharacterDataFromElement(line);
+                            logger.trace("attributeName: " + attributeName);
+
+                            NodeList deviceID = element.getElementsByTagName("value");
+                            line = (Element) deviceID.item(0);
+                            String attributeValue = getCharacterDataFromElement(line);
+                            logger.trace("attributeValue: " + attributeValue);
+
+                            switch (attributeName) {
+                                case "Mode":
+                                    State newMode = new StringType("Brewing");
+                                    switch (attributeValue) {
+                                        case "0":
+                                            updateState(CHANNEL_STATE, OnOffType.ON);
+                                            newMode = new StringType("Refill");
+                                            updateState(CHANNEL_COFFEEMODE, newMode);
+                                            break;
+                                        case "1":
+                                            updateState(CHANNEL_STATE, OnOffType.OFF);
+                                            newMode = new StringType("PlaceCarafe");
+                                            updateState(CHANNEL_COFFEEMODE, newMode);
+                                            break;
+                                        case "2":
+                                            updateState(CHANNEL_STATE, OnOffType.OFF);
+                                            newMode = new StringType("RefillWater");
+                                            updateState(CHANNEL_COFFEEMODE, newMode);
+                                            break;
+                                        case "3":
+                                            updateState(CHANNEL_STATE, OnOffType.OFF);
+                                            newMode = new StringType("Ready");
+                                            updateState(CHANNEL_COFFEEMODE, newMode);
+                                            break;
+                                        case "4":
+                                            updateState(CHANNEL_STATE, OnOffType.ON);
+                                            newMode = new StringType("Brewing");
+                                            updateState(CHANNEL_COFFEEMODE, newMode);
+                                            break;
+                                        case "5":
+                                            updateState(CHANNEL_STATE, OnOffType.OFF);
+                                            newMode = new StringType("Brewed");
+                                            updateState(CHANNEL_COFFEEMODE, newMode);
+                                            break;
+                                        case "6":
+                                            updateState(CHANNEL_STATE, OnOffType.OFF);
+                                            newMode = new StringType("CleaningBrewing");
+                                            updateState(CHANNEL_COFFEEMODE, newMode);
+                                            break;
+                                        case "7":
+                                            updateState(CHANNEL_STATE, OnOffType.OFF);
+                                            newMode = new StringType("CleaingSoaking");
+                                            updateState(CHANNEL_COFFEEMODE, newMode);
+                                            break;
+                                        case "8":
+                                            updateState(CHANNEL_STATE, OnOffType.OFF);
+                                            newMode = new StringType("BrewFailCarafeRemoved");
+                                            updateState(CHANNEL_COFFEEMODE, newMode);
+                                            break;
+                                    }
+                                    break;
+                                case "ModeTime":
+                                    if (attributeValue != null) {
+                                        State newAttributeValue = new DecimalType(attributeValue);
+                                        updateState(CHANNEL_MODETIME, newAttributeValue);
+                                    }
+                                    break;
+                                case "TimeRemaining":
+                                    if (attributeValue != null) {
+                                        State newAttributeValue = new DecimalType(attributeValue);
+                                        updateState(CHANNEL_TIMEREMAINING, newAttributeValue);
+                                    }
+                                    break;
+                                case "WaterLevelReached":
+                                    if (attributeValue != null) {
+                                        State newAttributeValue = new DecimalType(attributeValue);
+                                        updateState(CHANNEL_WATERLEVELREACHED, newAttributeValue);
+                                    }
+                                    break;
+                                case "CleanAdvise":
+                                    if (attributeValue != null) {
+                                        State newAttributeValue = attributeValue.equals("0") ? OnOffType.OFF
+                                                : OnOffType.ON;
+                                        updateState(CHANNEL_CLEANADVISE, newAttributeValue);
+                                    }
+                                    break;
+                                case "FilterAdvise":
+                                    if (attributeValue != null) {
+                                        State newAttributeValue = attributeValue.equals("0") ? OnOffType.OFF
+                                                : OnOffType.ON;
+                                        updateState(CHANNEL_FILTERADVISE, newAttributeValue);
+                                    }
+                                    break;
+                                case "Brewing":
+                                    if (attributeValue != null) {
+                                        State newAttributeValue = getDateTimeState(attributeValue);
+                                        if (newAttributeValue != null) {
+                                            updateState(CHANNEL_BREWING, newAttributeValue);
+                                        }
+                                    }
+                                    break;
+                                case "Brewed":
+                                    if (attributeValue != null) {
+                                        State newAttributeValue = getDateTimeState(attributeValue);
+                                        if (newAttributeValue != null) {
+                                            updateState(CHANNEL_BREWED, newAttributeValue);
+                                        }
+                                    }
+                                    break;
+                                case "Cleaning":
+                                    if (attributeValue != null) {
+                                        State newAttributeValue = new DecimalType(attributeValue);
+                                        updateState(CHANNEL_CLEANING, newAttributeValue);
+                                    }
+                                    break;
+                                case "LastCleaned":
+                                    if (attributeValue != null) {
+                                        State newAttributeValue = new DecimalType(attributeValue);
+                                        updateState(CHANNEL_LASTCLEANED, newAttributeValue);
+                                    }
+                                    break;
+                            }
+                        }
+                    } catch (Exception e) {
+                        logger.error("Failed to parse attributeList for WeMo CoffeMaker '{}'", this.getThing().getUID(),
+                                e);
+                    }
+                }
+            }
+        } catch (Exception e) {
+            logger.error("Failed to get attributes for device '{}'", getThing().getUID(), e);
+        }
+    }
+
+    public State getDateTimeState(String attributeValue) {
+        if (attributeValue != null) {
+            long value = 0;
+            try {
+                value = Long.parseLong(attributeValue) * 1000; // convert s to ms
+            } catch (NumberFormatException e) {
+                logger.error("Unable to parse attributeValue '{}' for device '{}'; expected long", attributeValue,
+                        getThing().getUID());
+                return null;
+            }
+            GregorianCalendar brewCal = new GregorianCalendar();
+            brewCal.setTimeInMillis(value);
+            State dateTimeState = new DateTimeType(brewCal);
+            if (dateTimeState != null) {
+                logger.trace("New attribute brewed '{}' received", dateTimeState);
+                return dateTimeState;
+            }
+        }
+        return null;
+    }
+
+    public String getWemoURL(String actionService) {
+        URL descriptorURL = service.getDescriptorURL(this);
+        String wemoURL = null;
+        if (descriptorURL != null) {
+            String deviceURL = StringUtils.substringBefore(descriptorURL.toString(), "/setup.xml");
+            wemoURL = deviceURL + "/upnp/control/" + actionService + "1";
+            return wemoURL;
+        }
+        return null;
+    }
+
+    public static String getCharacterDataFromElement(Element e) {
+        Node child = e.getFirstChild();
+        if (child instanceof CharacterData) {
+            CharacterData cd = (CharacterData) child;
+            return cd.getData();
+        }
+        return "?";
+    }
+
+    @Override
+    public void onStatusChanged(boolean status) {
+    }
+
+    @Override
+    public Collection<ThingUID> removeOlderResults(DiscoveryService source, long timestamp,
+            Collection<ThingTypeUID> thingTypeUIDs) {
+        return null;
+    }
+
+}

--- a/extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoCoffeeHandler.java
+++ b/extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoCoffeeHandler.java
@@ -108,7 +108,7 @@ public class WemoCoffeeHandler extends BaseThingHandler implements UpnpIOPartici
 
         super(thing);
 
-        logger.debug("Creating a WemoCoffeeHandler V0.3 for thing '{}'", getThing().getUID());
+        logger.debug("Creating a WemoCoffeeHandler V0.4 for thing '{}'", getThing().getUID());
 
         if (upnpIOService != null) {
             this.service = upnpIOService;
@@ -427,14 +427,6 @@ public class WemoCoffeeHandler extends BaseThingHandler implements UpnpIOPartici
                                         updateState(CHANNEL_FILTERADVISE, newAttributeValue);
                                     }
                                     break;
-                                case "Brewing":
-                                    if (attributeValue != null) {
-                                        State newAttributeValue = getDateTimeState(attributeValue);
-                                        if (newAttributeValue != null) {
-                                            updateState(CHANNEL_BREWING, newAttributeValue);
-                                        }
-                                    }
-                                    break;
                                 case "Brewed":
                                     if (attributeValue != null) {
                                         State newAttributeValue = getDateTimeState(attributeValue);
@@ -443,16 +435,12 @@ public class WemoCoffeeHandler extends BaseThingHandler implements UpnpIOPartici
                                         }
                                     }
                                     break;
-                                case "Cleaning":
-                                    if (attributeValue != null) {
-                                        State newAttributeValue = new DecimalType(attributeValue);
-                                        updateState(CHANNEL_CLEANING, newAttributeValue);
-                                    }
-                                    break;
                                 case "LastCleaned":
                                     if (attributeValue != null) {
-                                        State newAttributeValue = new DecimalType(attributeValue);
-                                        updateState(CHANNEL_LASTCLEANED, newAttributeValue);
+                                        State newAttributeValue = getDateTimeState(attributeValue);
+                                        if (newAttributeValue != null) {
+                                            updateState(CHANNEL_LASTCLEANED, newAttributeValue);
+                                        }
                                     }
                                     break;
                             }

--- a/extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoCoffeeHandler.java
+++ b/extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/handler/WemoCoffeeHandler.java
@@ -385,7 +385,7 @@ public class WemoCoffeeHandler extends BaseThingHandler implements UpnpIOPartici
                                             break;
                                         case "7":
                                             updateState(CHANNEL_STATE, OnOffType.OFF);
-                                            newMode = new StringType("CleaingSoaking");
+                                            newMode = new StringType("CleaningSoaking");
                                             updateState(CHANNEL_COFFEEMODE, newMode);
                                             break;
                                         case "8":

--- a/extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/internal/WemoHandlerFactory.java
+++ b/extensions/binding/org.eclipse.smarthome.binding.wemo/src/main/java/org/eclipse/smarthome/binding/wemo/internal/WemoHandlerFactory.java
@@ -17,6 +17,7 @@ import java.util.Set;
 import org.eclipse.smarthome.binding.wemo.WemoBindingConstants;
 import org.eclipse.smarthome.binding.wemo.discovery.WemoLinkDiscoveryService;
 import org.eclipse.smarthome.binding.wemo.handler.WemoBridgeHandler;
+import org.eclipse.smarthome.binding.wemo.handler.WemoCoffeeHandler;
 import org.eclipse.smarthome.binding.wemo.handler.WemoHandler;
 import org.eclipse.smarthome.binding.wemo.handler.WemoLightHandler;
 import org.eclipse.smarthome.binding.wemo.handler.WemoMakerHandler;
@@ -79,6 +80,10 @@ public class WemoHandlerFactory extends BaseThingHandlerFactory {
                 logger.debug("Creating a WemoHandler for thing '{}' with UDN '{}'", thing.getUID(),
                         thing.getConfiguration().get(UDN));
                 return new WemoHandler(thing, upnpIOService);
+            } else if (thingTypeUID.equals(WemoBindingConstants.THING_TYPE_COFFEE)) {
+                logger.debug("Creating a WemoCoffeeHandler for thing '{}' with UDN '{}'", thing.getUID(),
+                        thing.getConfiguration().get(UDN));
+                return new WemoCoffeeHandler(thing, upnpIOService);
             } else if (thingTypeUID.equals(WemoBindingConstants.THING_TYPE_MZ100)) {
                 return new WemoLightHandler(thing, upnpIOService);
             } else {

